File: 1658-minimum-operations-to-reduce-x-to-zero/main.cpp
class Solution {
public:
  int minOperations(vector<int>& nums, int x) {
    vector<int> left_sums, right_sums;
    left_sums.reserve(nums.size() + 1);
    right_sums.reserve(nums.size() + 1);
    
    left_sums.push_back(0);
    for (int i=0; i<nums.size() && left_sums.back() <= x; ++i) {
      left_sums.push_back(left_sums.back() + nums[i]);
    }
    
    right_sums.push_back(0);
    for (int i=nums.size()-1; i>-1 && right_sums.back() <= x; --i) {
      right_sums.push_back(right_sums.back() + nums[i]);
    }
    
    int op = nums.size() + 1;
    int right = right_sums.size() - 1;
    
    for (int left=0; left<left_sums.size(); ++left) {
      for (;right>-1;--right) {
        int sum = left_sums[left] + right_sums[right];
        if (sum <= x) {
          if (sum == x) {
            op = min(op, left + right);
          }
          break;
        }
      }
    }
    
    if (op <= nums.size()) {
      return op;
    }
    
    return -1;
  }
};

File: 1658-minimum-operations-to-reduce-x-to-zero/main.rs
impl Solution {
    pub fn min_operations(nums: Vec<i32>, x: i32) -> i32 {
        let mut left_sums = Vec::<i32>::with_capacity(nums.len());
        left_sums.push(0);
        for num in nums.iter() {
            left_sums.push(left_sums.last().unwrap() + num);
        }

        let mut right_sums = Vec::<i32>::with_capacity(nums.len());
        right_sums.push(0);
        for num in nums.iter().rev() {
            right_sums.push(right_sums.last().unwrap() + num);
        }

        let mut right = (right_sums.len() - 1) as i32;
        let mut op = (nums.len() + 1) as i32;

        left_sums.iter().enumerate().for_each(|(i, le)| {
            let left = i as i32;
            while right != -1 {
                let sum = le + right_sums[right as usize];
                if sum <= x {
                    if sum == x {
                        if op > left + right {
                            op = left + right;
                        }
                    }
                    break;
                }
                right -= 1;
            }
        });

        if (op <= nums.len() as i32) {
            return op;
        }

        return -1;
    }
}
